File: C++/quick_sort_step.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <limits>

using namespace std;

class TestQuickSortStep {
public:
    void main() {
        string input_filename;
        cout << "Enter CSV file name: ";  // Prompt user to input CSV filename
        cin >> input_filename;

        // Check if filename ends with ".csv"
        if (input_filename.size() < 4 || input_filename.substr(input_filename.size() - 4) != ".csv") {
            cerr << "Error: Please provide a valid CSV file." << endl; // If not CSV, print error
            return;
        }

        // Step 1: Load entire CSV file data into memory
        vector<pair<int, string>> full_data = read_data_from_csv(input_filename);
        if (full_data.empty()) {
            cerr << "Error: Could not read data from " << input_filename << ". Please ensure the file exists." << endl; // Error if file empty or unreadable
            return;
        }

        int start_idx, end_idx;
        while (true) {
            cout << "Enter start row: ";  // Ask user for start row (1-based)
            if (!(cin >> start_idx)) {   // Validate input is integer
                cout << "Invalid input. Please enter integers." << endl;
                cin.clear();              // Reset input stream error flags
                cin.ignore(numeric_limits<streamsize>::max(), '\n');  // Discard invalid input
                continue;                 // Retry input
            }

            cout << "Enter end row: ";   // Ask user for end row (1-based)
            if (!(cin >> end_idx)) {     // Validate input is integer
                cout << "Invalid input. Please enter integers." << endl;
                cin.clear();              // Reset input stream error flags
                cin.ignore(numeric_limits<streamsize>::max(), '\n');  // Discard invalid input
                continue;                 // Retry input
            }

            start_idx -= 1; // Convert to 0-based index for vector slicing
            end_idx -= 1;

            // Check if indices are valid and in correct order
            if (start_idx < 0 || end_idx >= full_data.size() || start_idx > end_idx) {
                cout << "Invalid row range. Enter values between 1 and " << full_data.size() << "." << endl;
                continue;                
            }
            break;                       
        }

        // Extract the subset of data between start_idx and end_idx (inclusive)
        vector<pair<int, string>> data_subset(full_data.begin() + start_idx, full_data.begin() + end_idx + 1);

        // Prepare output filename to record sorting steps
        string output_filename = "quick_sort_step_" + to_string(start_idx + 1) + "_" + to_string(end_idx + 1) + ".txt";
        ofstream output_file(output_filename);

        if (!output_file.is_open()) {  // Check if output file can be opened
            cerr << "Error opening file for writing: " << output_filename << endl;
            return;
        }

        // Write initial state of subset to output file (formatted)
        output_file << "[" << join(data_subset) << "]\n";

        // Call quick_sort on subset, passing output file for logging steps
        quick_sort(data_subset, 0, data_subset.size() - 1, output_file);

        cout << "Sorting steps written to " << output_filename << endl;  // Notify user
    }

private:
    vector<pair<int, string>> read_data_from_csv(const string& filename) {
        vector<pair<int, string>> data;
        ifstream file(filename);
        string line;
        int row_num = 0;

        if (!file.is_open()) {  // Check if file opens successfully
            cerr << "File '" << filename << "' not found." << endl;
            return data;         // Return empty vector on failure
        }

        // Read file line by line
        while (getline(file, line)) {
            row_num++;          // Keep track of line number for error messages
            stringstream ss(line);
            string num_str, word;
            if (getline(ss, num_str, ',') && getline(ss, word)) {  // Parse integer and string separated by comma
                try {
                    int number = stoi(num_str);  // Convert string to integer
                    data.emplace_back(number, word);  // Store tuple in vector
                } catch (...) {
                    cerr << "Skipping row " << row_num << " (invalid integer): " << line << endl; // Error on conversion
                }
            } else {
                cerr << "Skipping row " << row_num << " (malformed): " << line << endl;  // If row format incorrect
            }
        }

        return data;  // Return parsed data vector
    }

    string join(const vector<pair<int, string>>& data) {
        stringstream ss;
        // Format vector data into "int/string, int/string, ..." string
        for (size_t i = 0; i < data.size(); ++i) {
            ss << data[i].first << "/" << data[i].second;
            if (i < data.size() - 1) ss << ", ";
        }
        return ss.str();  // Return formatted string
    }

    void quick_sort(vector<pair<int, string>>& array, int low, int high, ofstream& log) {
        if (low < high) {
            // Partition array and get pivot index
            int pivot_index = partition(array, low, high, log);

            // Log pivot index and current state of array after partition
            log << "pi=" << pivot_index << " [" << join(array) << "]\n";

            // Recursively sort elements before pivot
            quick_sort(array, low, pivot_index - 1, log);

            // Recursively sort elements after pivot
            quick_sort(array, pivot_index + 1, high, log);
        }
    }

    int partition(vector<pair<int, string>>& array, int low, int high, ofstream& log) {
        int pivot = array[high].first;  // Choose last element as pivot
        int i = low - 1;                // Index of smaller element

        for (int j = low; j < high; ++j) {
            // If current element is less than or equal to pivot
            if (array[j].first <= pivot) {
                ++i;
                swap(array[i], array[j]);  // Swap elements
            }
        }

        swap(array[i + 1], array[high]);  // Place pivot in correct position
        return i + 1;                     // Return pivot index
    }
};

int main() {
    TestQuickSortStep sorter;
    sorter.main();
    return 0;
}

File: Python/quick_sort_step.py
import csv 

class TestQuickSortStep:
    def format_output_list(self, data_list):
        """
        Takes a list of (number, string) pairs and formats them into strings like "123/label".
        This is mainly used to make the log output easier to read.
        """
        return [f"{item[0]}/{item[1]}" for item in data_list]

    def main(self):
        # Ask the user to provide the CSV file name (should be in the same folder)
        input_filename = input("Enter CSV file name : ")    
        if not input_filename.endswith('.csv'):
            print("Error: Please provide a valid CSV file.")
            return

        # Load all the data from the CSV file
        full_data = self.read_data_from_csv(input_filename)
        if full_data is None:
            print(f"Error: Could not read data from {input_filename}. Please ensure the file exists.")
            return

        # Ask the user which rows they want to sort (inclusive range)
        while True:
            try:
                start_row_input = input(f"Enter start row : ")
                end_row_input = input(f"Enter end row : ")

                # Convert user input from 1-based to 0-based index
                start_idx = int(start_row_input) - 1
                end_idx = int(end_row_input) - 1

                # Make sure the given range is within the data limits
                if not (0 <= start_idx <= end_idx < len(full_data)):
                    print(f"Invalid row range. Enter values between 1 and {len(full_data)}.")
                    continue
                break
            except ValueError:
                print("Invalid input. Please enter integers.")

        # Grab just the portion of data that the user wants to sort
        data_subset = full_data[start_idx : end_idx + 1]

        # Set the filename where we'll log the sorting steps
        output_filename = f"quick_sort_step_{start_row_input}_{end_row_input}.txt"

        try:
            # Open the output file and begin logging
            with open(output_filename, 'w', newline='') as output_file:
                # Write the initial state of the data before sorting
                output_file.write(f"[{', '.join(self.format_output_list(data_subset))}]\n")

                # Perform quick sort and log every step
                self.quick_sort(data_subset, 0, len(data_subset) - 1, output_file)

            print(f"Sorting steps written to {output_filename}")
        except Exception as e:
            print(f"Error during sorting or writing: {e}")

    def read_data_from_csv(self, filename):
        """
        Reads data from the CSV file and stores it as a list of (number, string) pairs.
        If a row is badly formatted or contains invalid data, it's skipped with a warning.
        """
        data = []
        try:
            with open(filename, 'r', newline='') as csvfile:
                reader = csv.reader(csvfile)
                for row_num, row in enumerate(reader, 1):  # Start counting from row 1
                    if len(row) == 2:
                        try:
                            number = int(row[0].strip())   # Convert first value to integer
                            label = row[1].strip()         # Clean up the second value (string)
                            data.append((number, label))
                        except ValueError:
                            print(f"Skipping row {row_num} (invalid integer): {row}")
                    else:
                        print(f"Skipping row {row_num} (malformed): {row}")
            return data
        except FileNotFoundError:
            print(f"File '{filename}' not found.")
            return None
        except Exception as e:
            print(f"Error reading CSV: {e}")
            return None

    def quick_sort(self, array, low, high, output_file):
        """
        The main recursive quick sort function.
        Uses the last element as the pivot, splits the list around it,
        and logs each step to a file.
        """
        if low < high:
            # Rearrange the array and find out where the pivot ends up
            pivot_index = self.partition(array, low, high, output_file)

            # Log the pivot position and the state of the array after this step
            output_file.write(f"pi={pivot_index} [{', '.join(self.format_output_list(array))}]\n")

            # Now sort the left and right halves
            self.quick_sort(array, low, pivot_index - 1, output_file)
            self.quick_sort(array, pivot_index + 1, high, output_file)

    def partition(self, array, low, high, output_file):
        """
        Helper function to do the actual partitioning around the pivot.
        It puts smaller values before the pivot and larger ones after it.
        Returns the final index where the pivot ends up.
        """
        pivot = array[high][0]  # Use the number part of the last tuple as pivot
        i = low - 1  # Keeps track of where to swap next

        for j in range(low, high): #Loops through the array from index low to high - 1.
            if array[j][0] <= pivot: #Checks if the number part of the current element is less than or equal to the pivot value.
                i += 1               #Moves the boundary for smaller elements to the right.
                array[i], array[j] = array[j], array[i]  # Swap values

        # Finally, move the pivot to its correct position
        array[i + 1], array[high] = array[high], array[i + 1]
        return i + 1  # Return pivot's final index

# If the script is run directly, start the sorting process
if __name__ == "__main__":
    sorter = TestQuickSortStep()
    sorter.main()
